graph/pull_v2.go

```diff
@@ -0,0 +1,361 @@
+package graph
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/manifest"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/pkg/progressreader"
+	"github.com/docker/docker/pkg/streamformatter"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/docker/registry"
+	"github.com/docker/docker/trust"
+	"github.com/docker/docker/utils"
+	"github.com/docker/libtrust"
+)
+
+type v2Puller struct{ *TagStore }
+
+func (p *v2Puller) Pull(image, name, tag string, endpoint registry.APIEndpoint, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (fallback bool, err error) {
+	// TODO(tiborvass): was ReceiveTimeout
+	repo, err := NewV2Repository(name, endpoint, imagePullConfig.MetaHeaders, imagePullConfig.AuthConfig)
+	if err != nil {
+		logrus.Debugf("Error getting v2 registry: %v", err)
+		return true, err
+	}
+
+	if err := p.pullV2Repository(repo, image, tag, imagePullConfig, sf); err != nil {
+		if registry.ContinueOnError(err) {
+			logrus.Debugf("Error trying v2 registry: %v", err)
+			return true, err
+		}
+		return false, err
+	}
+	return false, nil
+}
+
+func (p *v2Puller) pullV2Repository(repo distribution.Repository, image, tag string, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (err error) {
+	var tags []string
+	taggedName := image
+	if len(tag) > 1 {
+		tags = []string{tag}
+		taggedName = utils.ImageReference(image, tag)
+	} else {
+		var err error
+		tags, err = repo.Manifests().Tags()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	c, err := p.poolAdd("pull", taggedName)
+	if err != nil {
+		if c != nil {
+			// Another pull of the same repository is already taking place; just wait for it to finish
+			sf.FormatStatus("", "Repository %s already being pulled by another client. Waiting.", image)
+			<-c
+			return nil
+		}
+		return err
+	}
+	defer p.poolRemove("pull", taggedName)
+
+	var pulledNew bool
+	for _, tag := range tags {
+		pulledNew, err = p.pullV2Tag(repo, image, tag, taggedName, imagePullConfig.OutStream, sf)
+		if err != nil {
+			return err
+		}
+	}
+
+	WriteStatus(taggedName, imagePullConfig.OutStream, sf, pulledNew)
```

Comment:
no but pulledNew has to be `or`-d with `||`, because if all but the last tag was downloaded, then the message will be not as expected. Thanks again for catching this